### ISSUE TITLE
TTS Tacotron 2, "Weak" Alignment, any suggestions?

### ISSUE THREAD

**Issue by Imtinan1996** [espnet/espnet#1360]
So i am trying to train Tacotron 2 on some custom dataset, i have a single speaker dataset, that is roughly around 11 hours. 

I have trained other implementations of tacotron 1 before, and on this one implementation the alignment learnt was very good.

ESPNet though for some reason learns alignment, but its a bit "weak" meaning at every timestamp the predicted phonemes are a bit off sometimes.  

I am training this on 6 GPUs with batch size 32. I trained libritts as well from the default recipes, and according the the config, the model learnt very good alignment in only 30 epochs. But as can be seen from the GIF below, the alignment does get better, but it takes 800 epochs, and still its somewhat weak.

Can anyone give suggestions on what the problem could be, or what i could do to make things better, any help would be greatly appreciated.

EDIT:::

To not mess with the specifications alot, i have only changed the sampling rate in the config to match my datas SR, all other parameters like n_mels, nfft, etc remain the same. Could this be an issue? Should i resample my data to 24000 to match libritts specifications and try training the model again? My params are as follows

```
fs=16000      # sampling frequency
fmax=""       # maximum frequency
fmin=""       # minimum frequency
n_mels=80     # number of mel basis
n_fft=1024    # number of fft points
n_shift=256   # number of shift points
win_length="" # window length
 
```

Additionally this is what my melspectograms look like from feats.scp and feats.ark 

random feats.scp
![feat.scp](https://user-images.githubusercontent.com/30050508/68930536-0780c080-07a8-11ea-8709-32d8d10a690c.png)

random feats.ark
![feats.ark](https://user-images.githubusercontent.com/30050508/68930697-5dedff00-07a8-11ea-9d85-12f1a6d79dbf.png)

This is very different from what libritts features looked like, but my limited knowledge in signal processing does not help me identify what the problem could be

![Alignment spochs 280-870](https://user-images.githubusercontent.com/30050508/68928650-077ec180-07a4-11ea-886b-1babfc37106d.gif)

**Comment by sw005320** [comment]
My suggestion is to use a single GPU. We may not need to use 6 GPUs for such small data. ESPnet multiple GPU training is based on a simple data parallelization, and your method corresponds to the use of 6x32=192 batch size, which is too large for this data. Then, according to the general online-training theory, you require a lot of iterations to get some reasonable convergence.

**Comment by kan-bayashi** [comment]
Hi @Imtinan1996. Thank you for using ESPnet-TTS!
I will list some basic techniques to get clear attention
- Remove silence at the beginning and the end of utterances. You can follow the following part:
https://github.com/espnet/espnet/blob/5d2d2fb2be3a714c4b7b8e5dfb6135abca78e973/egs/m_ailabs/tts1/run.sh#L98-L107
- Use small batch-size for the small dataset as @sw005320 said. In our implementation, the batch size will be automatically multiplied by the number of GPUs.
- Make sure your input text is appropriately cleaned and normalized. (Number, abbreviation, case, and so on.)
- If your model is a single speaker, please use `ljspeech` configuration (e.g. https://github.com/espnet/espnet/blob/master/egs/ljspeech/tts1/conf/tuning/train_pytorch_tacotron2.v3.yaml). v3 applies guided attention loss to force the attention weight to be diagonal. Just changing the sampling rate might be OK.
- Why don't you limit the Mel range? Recently we use 80-7600 as default.

**Comment by Imtinan1996** [comment]
Thank you @sw005320 for the suggestions, i changed it to single-GPU and these are the results so far.

![single_gpu_train_GIF](https://user-images.githubusercontent.com/30050508/69005025-15178100-0935-11ea-9034-949449f425d6.gif)

As you can see, even after 200 epochs, the results are not so good.

@kan-bayashi , thank you so much for your suggestions, though i am already following some of them, for instance

- Silence trimming: i had removed this, but i have now added it to the recipe
- Smaller batches: As you can see above, i did go from multi to single GPU as per @sw005320 's suggestion, but it did not do much
- Text normalization: Yes, all input text has been appropriately cleaned and normalized manually, though i didn't personally check all the files i can state that the text is clean and normalized due to the fact that tacotron1 was trained on this with reasonably good outputs (barring naturalness) 
- ljspeech config: I was using just libritts recipe as reference, this is the config i am using for training, and as per the config, i have also removed the x-vector extraction part from the recipe, since that step was now useless.
- Mel range: Since i was using libritts as reference, it did not have specified mel ranges, so i let it be as such. I have now set them as advised.

In summary, apart from mel range limiting and silence trimming, i have already done everything suggested. I will retrain the model using these changes. Though i dont see how they could bring such a marginal change in performance, regardless i will do these changes, and get back to you guys.

Thank you both for the suggestions, they are highly appreciated.

**Comment by kan-bayashi** [comment]
@Imtinan1996 Sorry, I could not catch up your meaning. Did you already use this config?
https://github.com/espnet/espnet/blob/master/egs/ljspeech/tts1/conf/tuning/train_pytorch_tacotron2.v3.yaml
If so, v2 config is also worthwhile to use which utilize forward attention instead of location sensitive attention.

In your previous tacotron experiment, did you use reduction factor? Reduction factor also helps to learn diagonal attention.

And from your fbank plot, there are clear silence parts. In the input text, is there any corresponding character or symbol?

**Comment by kan-bayashi** [comment]
I remembered that I also met such weak attention in m-ailabs recipes, where the SNR is not so good. If the reason of the weak attention is derived from the SNR of the target speech, we should try another extension such as VAE-based Tacotron 2 but I’m not sure.

**Comment by Imtinan1996** [comment]
@kan-bayashi  yes i am already using https://github.com/espnet/espnet/blob/master/egs/ljspeech/tts1/conf/tuning/train_pytorch_tacotron2.v3.yaml, i will also try experimenting with V2 config aswell.

My previous experiments were on the implementation by Kyubyong here https://github.com/Kyubyong/tacotron, yes they did use a reduction factor of 5.

About the silence plots, yes they are existent in a large number of the audio files, but none exceed 250ms, and no they are not represented in the text, do you think i should write a script that trims these mid silences?

As far as the SNR goes, the recording were done in a studio, and there is next to no noise whatsoever, so i doubt that the weak attention is because of the SNR.

Just for reference, this is my attention plot from the aforementioned taco1 implementation
![alignment_356k](https://user-images.githubusercontent.com/30050508/69005442-e51fac00-093b-11ea-9bc1-53ee335a79ef.png)

**Comment by Imtinan1996** [comment]
UPDATE:

This is the alignment  after 125 epochs, its too early to make a blatant statement but the progress looks healthy, i will be sure to keep you guys updated

![espnet_alignment](https://user-images.githubusercontent.com/30050508/69032851-efeb4700-09f6-11ea-8bdc-9ea28a64be67.gif)

**Comment by kan-bayashi** [comment]
@Imtinan1996 It looks nice. What is the main defference?

**Comment by Imtinan1996** [comment]
i think the main difference was the silence trimming, since the aforementioned repo also utilizes silence trimming. 

Thank you both for all your help.

Also, can we please keep the issue open for now, i will close it once i am able to post the final results.

**Comment by Imtinan1996** [comment]
hi @kan-bayashi i needed some advice, so after further training the alignment is a lot better, but intelligibility in some samples is still an issue, _**especially**_ at the end of the sentences. 

During my experimentations with libritts i identified that the data json files do not consist of a stop token (something tacotron2 reportedly generates). So when i wrote my custom data2json, i did not append a stop token to the data. 

I also identified this in the data-jsons as well, that the output shape of the transciptions is 

> "shape": [ x,78 ],

Where 78 should be the size of the vocab. Now my question is that the vocab file for librispeech (train_clean_460_units.txt) shows symbols ranging from indexes 1-76, as per the documentation "0" is used for padding, so the total len of the vocab should be 77, i assumed that the 78th element would be the "stop token" that is automatically added during data loading. is it so? if not then why the extra one index in vocab size.

But now that the model is having difficulty in ending the sentence, there is a stutter of sorts at the end. i am starting to think my initial assumptions were wrong and that i should append a stop token. Should i modify my data2json and manually inject a stop token at the end of the tokenIDs? or do you think there is some other problem?

**Comment by kan-bayashi** [comment]
In that case 78th element is stop token, which is  automatically added during batch preparation.
The problem is that the generation does not stop appropriately, right? (Consequently, the word around the end of sentence will be repeated.) If so, I recommend adjusting the bce-pos-weight. Making it larger value helps to stop the generation correctly.

**Comment by kan-bayashi** [comment]
You can check the stop token probability in the inference from the figures in the probs directory.

**Comment by Imtinan1996** [comment]
> The problem is that the generation does not stop appropriately, right? (Consequently, the word around the end of the sentence will be repeated.)

Yes, this is exactly the problem.

Alright i will make the changes, but in accordance to that i have 2 questions

- What do you mean by a larger value? could you please give a rough estimate of what could be good to use, for example 2 is a larger value, and so is 10, and 100 and so on. What do you think would be large enough.
- Secondly, can i resume my training with this changed parameter? or is it better to start fresh?

**Comment by kan-bayashi** [comment]
> What do you mean by a larger value? 

Around 10 is enough. For example, we use 5.0 for Transformer.

> Secondly, can I resume my training with this changed parameter? or is it better to start fresh?

I'm not sure because I've never tried to change the hyperparameter during training. If you have time, it is better to train from scratch.

**Comment by kan-bayashi** [comment]
@Imtinan1996 Another solution is adjusting `threshold` in `decode.conf` but you should check the sequence of the stop probabilities in `probs` to do it.

**Comment by Imtinan1996** [comment]
Hi @kan-bayashi 

Upon running a few more experiments, i came across variating results (both good and bad), but some of these i couldn't explain, could you help me understand why such differences?

for example, this is the alignment on epoch 120, training on config V3 on a single GPU (batch size per GPU = 32)

![0006a1bc-db20-45a4-b1fc-c4d766ada191 ep 120](https://user-images.githubusercontent.com/30050508/69490538-0edc5400-0ea3-11ea-8ee9-4683f4e314db.png)

and this is the alignment on the exact same example with the same config but on 4 GPUs (batch size per GPU = 8)

Why is that the alignment so quick to learn on the second experiment?

NOTE: There was a small difference in both configs, the variable 'bce-pos-weight' in the distributed training was set to 1, whereas in the single GPU training was set to 2. I doubt that could have caused such a drastic change, but regardless i have started an experiment with the variable adjusted to 1 on single GPU training, lets see if we can observe the same results as exp 1

**Comment by kan-bayashi** [comment]
@Imtinan1996 Sorry for the late reply. Let me confirm one point. Did you explicitly reduce batchsize in config in the case of multi gpu?

**Comment by Imtinan1996** [comment]
@kan-bayashi yes i changed the batch-size to 8, and set ngpus to 4. Technically the effective batch size is still 32, but yet i see this difference

**Comment by kan-bayashi** [comment]
I see. That's is strange.
I have no idea why the speed is different.
How about the final results? Is it different too much?

**Comment by Imtinan1996** [comment]
Yes, it is very different, without the distributed smaller batch, the alignment almost never gets better (tested until ~1000 epochs)

**Comment by kan-bayashi** [comment]
Hmm... I'm not sure why it has happened.
How about the quality of the generated speech?

**Comment by Imtinan1996** [comment]
The quality for the distributed training is desirable, whereas in the single GPU training some examples are okay(with a few mistakes here and there) and some are hardly intelligible. 

Do you think in terms of gradient updation there is some bug, that is non existent in gradient averaging / batch update?

**Comment by kan-bayashi** [comment]
OK. I will check the code in detail.

**Comment by kan-bayashi** [comment]
I came up with one possible reason.
This line is the loss calculation.
https://github.com/espnet/espnet/blob/448947bf5a93b03208bfda5e8acb5428b3ac93da/espnet/tts/pytorch_backend/tts.py#L150
In the case of single gpu, `self.model(**x)` is the scalar value of loss.
The loss is calculated as the average of all of the bins of outputs.
Therefore, the long sequence in the batch is dominant.

However, in the case of multi-gpu, `self.model(**x)` is the vector of loss values whose size is `(#gpus,)`. 
Each item in the vector is the loss calculated as the average of the split batch.
The items in the batch are sorted by the length, so the number of output bins in gpu 1 is bigger than that in gpu 4 and therefore, the number of bins is different among the gpus.
Consequently, if we average the loss vector, it is similar to the length normalized loss.

So I think, this phenomenon suggests that the length-normalized loss is effective for the TTS models. 
Here the length-normalized loss means: 
loss `(#batch, #length, #dim)` -> normalize by length -> `(#batch, #dim)` -> normalize by remaining bins -> loss scalar

Currently, we perform normalization over all of the bins:
loss `(#batch, #length, #dim)` -> normalize over bins -> loss scalar

**Comment by kan-bayashi** [comment]
By default, we use `batch-sort-key: shuffle`.
So the length of each item in the batch is different.
https://github.com/espnet/espnet/blob/448947bf5a93b03208bfda5e8acb5428b3ac93da/egs/ljspeech/tts1/conf/tuning/train_pytorch_tacotron2.v3.yaml#L39
If we use `batch-sort-key: output`, this issue will be solved.
But I'm not sure which is better length-sorted batch or `shuffle`+length-normalized loss.

**Comment by Imtinan1996** [comment]
Alright, i will try running an experiment with `length-sorted batch` in comparison to `shuffle+length-normalized loss` (which i already have in training). 

And i will share the results whenever i can.

**Comment by Imtinan1996** [comment]
Also there is one other concern that i have, i havent seen the implementation in depth, but i assume that the number of steps per epoch is derived from the batch size and the data length `(n_steps = len(data) / batch_size )` If this is so, then i fear my comparison might be unfair.

Lets assume i have 3200 samples

When i set the batch size to 32 on a single GPU, the effective batch size is 32 and number of steps per epoch is 3200/32 = 1000

When i set the batch size to 8 on 4 gpus, the effective batchsize is 8*4 = 32, but this time the number of steps per epoch would be 3200/8 = 4000.

This means that multi GPU training sees alot more examples in an epoch than single GPU training does. 

Hence a much fairer comparison would be comparing alignment for multi GPU training at epoch 50, and single GPU training at epoch 200. 

Do you think my comparison was unfair?

**Comment by kan-bayashi** [comment]
In the case of multi-gpu, first, we make the list of mini-batches with the batch_size (= batch_size * num_gpus). And then each mini-batch will be split into the number of gpus in `forward` calculation. Therefore, the number of steps will be the same between `batch_size=32 & ngpu=1` and `batch_size=8 & ngpu=4`.

**Comment by kan-bayashi** [comment]
> shuffle+length-normalized loss (which i already have in training).

Note that current implementation is not `length-normalized`.
The current implementation is just average over all of the bins.

**Comment by kan-bayashi** [comment]
Sorry, `length-normalized` may confuse.
I mean `length-weighted-normalized` (not sure...).
```python
[ins] In [1]: x1 = np.arange(5)
[ins] In [2]: x2 = np.arange(4)
# now
[ins] In [3]: length_normalized = (x1.sum() + x2.sum()) / (len(x1) + len(x2))
# maybe better?
[ins] In [4]: length_weighted_normalized = (x1.mean() + x2.mean()) / 2
```

**Comment by kan-bayashi** [comment]
I made new option `use-weighted-masking` in #1397.
I will also compare the results with ljspeech recipe.

**Comment by kan-bayashi** [comment]
I added the results in #1397.
In the case of LJspeech, the results are almost the same.

**Comment by Imtinan1996** [comment]
hi @kan-bayashi thank you for making the changes. i was on a break so i am getting back a bit late. I will pull and try said experiments. Thanks alot
